qutip/nonmarkov/bofin.py

```diff
@@ -0,0 +1,1553 @@
+
+        if (
+            type(ckAR[0]) == list
+            or type(vkAR[0]) == list
+            or type(ckAR[0]) == list
+            or type(ckAI[0]) == list
+        ):
+            raise RuntimeError(
+                "Lists of coefficients should be one dimensional."
+            )
+
+        if len(ckAR) != len(vkAR) or len(ckAI) != len(vkAI):
+            raise RuntimeError(
+                "Spectral density correlation coefficients not "
+                + "specified correctly."
+            )
+
+        # Check for close vk's.
+        # Just gives warning, and continues.
+        for i in range(len(vkAR)):
+            for j in range(i + 1, len(vkAR)):
+                if np.isclose(vkAR[i], vkAR[j], rtol=1e-5, atol=1e-7):
+                    warnings.warn(
+                        "Expected simplified input. "
+                        "Consider collating equal frequency parameters."
+                    )
+
+        for i in range(len(vkAI)):
+            for j in range(i + 1, len(vkAI)):
+                if np.isclose(vkAI[i], vkAI[j], rtol=1e-5, atol=1e-7):
+                    warnings.warn(
+                        "Expected simplified input.  "
+                        "Consider collating equal frequency parameters."
+                    )
+
+        if type(H_sys) == list:
+            self.H_sys = QobjEvo(H_sys)
+        else:
+            self.H_sys = H_sys
+
+        nr = len(ckAR)
+        ni = len(ckAI)
+        ckAR = list(ckAR)
+        ckAI = list(ckAI)
+        vkAR = list(vkAR)
+        vkAI = list(vkAI)
+        coup_op = deepcopy(coup_op)
+
+        # Check to make list of coupling operators
+
+        if type(coup_op) != list:
+            coup_op = [coup_op for i in range(nr + ni)]
+
+        # Check for handling the case where real and imaginary exponents
+        # are close.
+        # This happens in the normal overdamped drude-lorentz case.
+        # We give a warning to tell the user this collation is being done
+        # automatically.
+        common_ck = []
+        real_indices = []
+        common_vk = []
+        img_indices = []
+        common_coup_op = []
+        for i in range(len(vkAR)):
+            for j in range(len(vkAI)):
+                if np.isclose(
+                    vkAR[i], vkAI[j], rtol=1e-5, atol=1e-7) and np.allclose(
+                    coup_op[i], coup_op[nr + j], rtol=1e-5, atol=1e-7
+                ):
+                    warnings.warn(
+                        "Two similar real and imag exponents have been "
+                        "collated automatically."
+                    )
+                    common_ck.append(ckAR[i])
+                    common_ck.append(ckAI[j])
+                    common_vk.append(vkAR[i])
+                    common_vk.append(vkAI[j])
+                    real_indices.append(i)
+                    img_indices.append(j)
+                    common_coup_op.append(coup_op[i])
+
```

Comment:
the above checks if two exponents are similar, and collates them into a single exponent. This is pretty important for efficiency and convergence in some cases, particularly the Drude-Loretnz spectral density, but its a little clunky.  I think at least we need a better warning message.   E.g., since these 'new' collated exponents are added to the end of the list, if the user is trying to do something with the ADOs, they may get mixed up about how various ADOs correspond to various exponents, or various baths, since the ADOs ordering will be inconsistent with the input lists of parameters.  Perhaps returning the ADOs from run() or steady_state() with a dictionary showing which ADOs correspond to which input parameters/baths/etc is an option?   This is a pretty niche issue, and non of the examples we put together ran into it, it came up in an email from a user (Hai-Feng Yang)